modules/supervisord.nix

```diff
@@ -0,0 +1,186 @@
+{ pkgs, lib, config, ... }:
+let
+  inherit (lib) types;
+
+  cfg = config.supervisord;
+
+  format = pkgs.formats.ini {};
+
+  programType = types.submodule ({ name, config, ... }: {
+    options = {
+      enable = lib.mkOption {
+        description = lib.mdDoc ''
+          Whether to enable this program.
+        '';
+        type = types.bool;
+        default = true;
+      };
+      command = lib.mkOption {
+        description = lib.mdDoc ''
+          The command that will be run as the service's main process.
+        '';
+        type = types.str;
+      };
+      script = lib.mkOption {
+        description = lib.mdDoc ''
+          Shell commands executed as the service's main process.
+        '';
+        type = types.lines;
+        default = "";
+      };
+      path = lib.mkOption {
+        description = lib.mdDoc ''
+          Packages added to the service's PATH environment variable.
+        '';
+        type = types.listOf (types.either types.package types.str);
+        default = [];
+      };
+      autoRestart = lib.mkOption {
+        description = lib.mdDoc ''
+          Whether to automatically restart the process if it exits.
+
+          If `unexpected`, the process will be restarted if it exits
+          with an exit code not listed in the programs's `exitcodes`
+          configuration.
+        '';
+        type = types.either types.bool (types.enum [ "false" "true" "unexpected" ]);
+        default = "unexpected";
+      };
+      environment = lib.mkOption {
+        description = lib.mdDoc ''
+          Environment variables passed to the service's process.
+        '';
+        type = types.attrsOf types.str;
+        default = {};
+      };
+      extraConfig = lib.mkOption {
+        description = lib.mdDoc ''
+          Extra structured configurations to add to the [program:x] section.
+        '';
+        type = types.attrsOf types.str;
+        default = {};
+      };
+    };
+    config = {
+      command = lib.mkIf (config.script != "")
+        (toString (pkgs.writeShellScript "${name}-script.sh" config.script));
+
+      environment.PATH = lib.mkDefault (lib.makeBinPath config.path);
+    };
+  });
+
+  renderAtom = val:
+    if builtins.isBool val then if val then "true" else "false"
+    else toString val;
+
+  renderProgram = program: let
+    section = {
+      inherit (program) command;
+      autorestart = program.autoRestart;
```

Comment:
I think it would also make sense to define `autostart` as an option, sounds like a common option.

Reply:
Not really. For example I want to enable sshd (to be able to start and stop it on demand) but not enabling autostart

Reply:
Ok, added (both supervisord and sshd).